### PR TITLE
update attach to use external objs

### PR DESCRIPTION
**Pull request by juanvallejo**
**Release note**:
```release-note
NONE
```

Updates attach command to use external versions.

cc @deads2k @soltysh

### CODE REVIEW

File: pkg/kubectl/cmd/attach.go [hunk 1]
```diff
@@ -111,146 +156,115 @@ func (*DefaultRemoteAttach) Attach(method string, url *url.URL, config *restclie
...
 	return nil
 }
 
 // Validate checks that the provided attach options are specified.
-func (p *AttachOptions) Validate() error {
-	allErrs := []error{}
-	if len(p.PodName) == 0 {
-		allErrs = append(allErrs, errors.New("pod name must be specified"))
+func (o *AttachOptions) Validate(args []string) error {
+	if len(args) == 0 {
```
Comment (soltysh) [inline pkg/kubectl/cmd/attach.go]: You can use `o.Resources`, and don't pass `args` here at all. Maybe renaming then to `Args` would make it more obvious.

File: pkg/kubectl/cmd/attach.go [hunk 2]
```diff
@@ -265,75 +279,63 @@ func (p *AttachOptions) Run() error {
 			sizeQueue = t.MonitorSize(&sizePlusOne, size)
 		}
 
-		// unset p.Err if it was previously set because both stdout and stderr go over p.Out when tty is
+		// unset o.Err if it was previously set because both stdout and stderr go over o.Out when tty is
 		// true
-		p.ErrOut = nil
+		o.ErrOut = nil
```
Comment (soltysh) [inline pkg/kubectl/cmd/attach.go]: I'm hesitant about this action. We're modifying global `IOStreams` object. I'd prefer we make a local copy of `ErrOut` and pass that to `AttachFunc` as needed.

File: pkg/kubectl/cmd/run.go [hunk 3]
```diff
@@ -22,18 +22,20 @@ import (
 	"github.com/docker/distribution/reference"
 	"github.com/spf13/cobra"
 
+	"github.com/golang/glog"
 	"k8s.io/client-go/dynamic"
```
Comment (soltysh) [inline pkg/kubectl/cmd/run.go]: Move this import to all k8s imports.

File: pkg/kubectl/polymorphichelpers/attachablepodforobject.go [hunk 4]
```diff
@@ -24,30 +24,29 @@ import (
 	"k8s.io/api/core/v1"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	coreclient "k8s.io/client-go/kubernetes/typed/core/v1"
```
Comment (soltysh) [inline pkg/kubectl/polymorphichelpers/attachablepodforobject.go]: `corev1client` is the preferred alias.

File: pkg/kubectl/polymorphichelpers/helpers.go [hunk 5]
```diff
@@ -52,15 +51,11 @@ func GetFirstPod(client coreclient.PodsGetter, namespace string, selector string
 	pods := []*v1.Pod{}
 	for i := range podList.Items {
 		pod := podList.Items[i]
-		externalPod := &v1.Pod{}
```
Comment (soltysh) [inline pkg/kubectl/polymorphichelpers/helpers.go]: One more down :+1:

File: pkg/kubectl/cmd/logs.go [hunk 6]
```diff
@@ -228,16 +251,37 @@ func (o *LogsOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, args []str
 	return nil
 }
 
-func (o LogsOptions) Validate() error {
-	logsOptions, ok := o.Options.(*api.PodLogOptions)
+func (o LogsOptions) Validate(args []string) error {
```
Comment (soltysh) [inline pkg/kubectl/cmd/logs.go]: If you need pass `args` through `LogsOptions` struct. `Validate` should not accept input parameters.

**Review by soltysh — CHANGES_REQUESTED** [review]
I left you some comments, also please squash that last commit into the first one.

**Comment by juanvallejo** [comment]
@soltysh thanks, comments addressed

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Pull request merged**